GET /api/user/me 테스트 작성

youngxpepp opened the pull request:
# GET /api/user/me 테스트 코드 작성

## 📎 해당 이슈

이슈 링크 ( #25 )

## 🛠 구현 내용 

- UserService.getUserById() 테스트 작성
- GET /api/user/me 테스트 작성

## 🙋‍ 리뷰어 참고 사항 

- 각 단위 테스트 간의 어떻게 독립적인 데이터베이스 환경을 갖는지 보시면 좋을 것 같아요!

youngxpepp committed:
[TEST] UserService.getUserById() 테스트 작성

- id 기반으로 사용자를 조회하는 메서드의 테스트 작성
- 테스트에서의 트랜잭션을 관리하는 TestTransactionDelegate 추가
- TestTransactionDelegate를 통해서 각 테스트의 독립적인 데이터베이스 환경을 갖음
- ESLint의 dangling 옵션 해제

youngxpepp committed:
[TEST] GET /api/user/me 테스트 작성

Park-SooYeon commented on server/test/service/UserService.test.js:
```diff
@@ -0,0 +1,51 @@
+
+    test('UserService.getUserById(): user0을 조회했을 때 user0 반환', async () => {
+        const userService = UserService.getInstance();
+        await TestTransactionDelegate.transaction(async () => {
+            // given
+            const entityManager = getEntityManagerOrTransactionManager();
```
transaction 함수 내부에서 선언된 entityManager 변수는 하나의 트랜잭션을 만들기 위해서 사용한 엔티티 매니저이고, callback 함수에서 선언된 entityManager 변수는 쿼리 수행을 위해 사용한 엔티티 매니저인걸까요.🤔
그렇다면 서비스함수를 작성할 때 사용한 @Transactional()과 getRepository를 사용해서 트랜젝션과 쿼리 수행을 하는 것도 결국은 동일할 것이라고 생각이 되는데 테스트 코드에서는 왜 @Transactional()과 getRepository를 사용하지 않고 엔티티 매니저를 가져와서 사용하는지 궁금합니다.
혹시 차이가 존재하는걸까요?👀

youngxpepp replied:
TestTransactionDelegate.transaction() 내부에 선언된 EntityManager는 트랜잭션을 만들기 위함이 아니고, 롤백을 위함입니다. Raw query 를 사용하려면 해당 트랜잭션에 접근하여 쿼리를 날려줄 주체가 필요하거든요ㅎㅎ  callback 내부에 선언된 EntityManager는 많은 Repository를 쓰지 않기 위해서 사용했습니다. Repository 내부 동작을 살펴보면 EntityManager가 수행합니다.
이 부분에서 현재 에러가 나서 해결하는 중입니다. 곧 커밋 할게요~

Park-SooYeon commented on server/test/service/UserService.test.js:
```diff
@@ -0,0 +1,51 @@
+        const userService = UserService.getInstance();
+        await TestTransactionDelegate.transaction(async () => {
+            try {
+                await userService.getUserById(0);
+                fail();
```
유저를 찾을 수 없다면 getUserById 함수가 실행되는 도중에 catch 문으로 이동이 되어 getUserById 함수의 아래 코드는 실행되지 않을 것 같은데 fail 함수를 사용해주는 이유가 무엇인가요?👀

youngxpepp replied:
테스트 실패 케이스를 만들어 주기 위함이에요. 에러가 발생하지 않는다면 테스트는 자연스레 실패하게 되겠죵?

Park-SooYeon replied:
에러가 발생해야하는데 발생하지 않는 경우에는 테스트를 실패하게 만드는 것이 맞겠군요.
함수가 에러를 발생시켜야 하는 상황의 테스트 코드를 작성해보지 않아서 거기까지는 생각 못했네요!
잘 배워갑니다. 감사합니다!!🙌

Park-SooYeon reviewed (commented):
수고하셨습니다.
supertest에서 요청 생성을 할때, request가 아니라 agent를 사용할수도 있군요!
typeORM에서 EntityManager를 제공하는 것도 처음 알았습니다..항상 많이 배워갑니다!👍

youngxpepp committed:
[FIX] 테스트 트랜잭션 에러 해결

- Transactional 내부에서 EntityManager를 불러올 때 Connection 이름인 'default'를 인자로 보냄
- typeorm-transactional-cls-hooked 버전 업데이트

Park-SooYeon reviewed (approved):
버전에 따라 사용법이 변경되서 생긴 오류였을까요?👀
고생하셨습니다!

dh00n reviewed (approved):
오 롤백 !
참고해서 테스트코드 작성하겠습니다

youngxpepp committed:
Merge branch 'develop/web' into feature/get-user-test
